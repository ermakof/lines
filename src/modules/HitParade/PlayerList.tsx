import React, { FC, memo } from 'react';
import { TUserList } from '@src/modules/HitParade/models';
import Grid from '@src/layout/Grid';
import { sortByScore } from '@src/modules/HitParade/utils';
import PlayerListHeader from '@src/modules/HitParade/PlayerListHeader';
import PlayerListRow from '@src/modules/HitParade/PlayerListRow';

interface IPlayerList {
  list: TUserList;
  loginTime?: number;
}
const PlayerList: FC<IPlayerList> = ({ list, loginTime }) => {
  const userList = Object.values(list).sort(sortByScore);
  return (
    <Grid role="grid">
      <PlayerListHeader />
      {userList.map((user, ind) => {
        return (
          <PlayerListRow
            key={user.ts}
            user={user}
            winner={ind + 1}
            highlighted={user.ts === loginTime}
          />
        );
      })}
    </Grid>
  );
};

export default memo(PlayerList);
