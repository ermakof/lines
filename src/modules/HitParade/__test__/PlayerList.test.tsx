import React from 'react';
import { render, screen } from '@testing-library/react';

import PlayerList from '@src/modules/HitParade/PlayerList';

const playerList = {
  1646851461537: { login: 'Игрок 1', ts: 1646851461537, score: 100 },
  1646855489093: { login: 'Игрок 2', ts: 1646855489093, score: 200 },
  1646859512749: { login: 'Игрок 3', ts: 1646859512749, score: 300 },
};

describe('PlayerList', () => {
  it('Render snapshot', () => {
    const { asFragment } = render(<PlayerList list={playerList} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Render form and grid', () => {
    render(<PlayerList list={playerList} />);
    const grid = screen.getByRole(/grid/gi);
    expect(grid).toBeInTheDocument();
    const header = screen.getByRole(/^header$/gi);
    expect(header).toBeInTheDocument();
    const rows = screen.getAllByRole(/^row$/gi);
    expect(rows).toHaveLength(3);
  });

  it('Render header values', () => {
    render(<PlayerList list={playerList} />);
    const headerTime = screen.getByText(/Время/gi);
    expect(headerTime).toBeInTheDocument();
    const headerLogin = screen.getByText(/Имя/gi);
    expect(headerLogin).toBeInTheDocument();
    const headerScore = screen.getByText(/Счет/gi);
    expect(headerScore).toBeInTheDocument();
    const headerWin = screen.getByText(/Место/gi);
    expect(headerWin).toBeInTheDocument();
  });

  it('Render grid values', () => {
    render(<PlayerList list={playerList} />);
    const player1 = screen.getByText(/Игрок 1/gi);
    expect(player1).toBeInTheDocument();
    const player2 = screen.getByText(/Игрок 2/gi);
    expect(player2).toBeInTheDocument();
    const player3 = screen.getByText(/Игрок 3/gi);
    expect(player3).toBeInTheDocument();
    const score1 = screen.getByText(/100/gi);
    expect(score1).toBeInTheDocument();
    const score2 = screen.getByText(/200/gi);
    expect(score2).toBeInTheDocument();
    const score3 = screen.getByText(/300/gi);
    expect(score3).toBeInTheDocument();
    const time1 = screen.getByText(/09.03.2022, 23:58:32/gi);
    expect(time1).toBeInTheDocument();
    const time2 = screen.getByText(/09.03.2022, 22:51:29/gi);
    expect(time2).toBeInTheDocument();
    const time3 = screen.getByText(/09.03.2022, 21:44:21/gi);
    expect(time3).toBeInTheDocument();
  });
});
