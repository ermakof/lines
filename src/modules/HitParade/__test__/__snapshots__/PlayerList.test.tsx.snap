// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerList Render snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-11wxzwa"
    role="grid"
  >
    <div
      class="css-8eb8xz"
      role="header"
    >
      <div
        class="css-1hygdxi"
        role="col"
      >
        Время
      </div>
      <div
        class="css-182swot"
        role="col"
      >
        Имя
      </div>
      <div
        class="css-lbgr5s"
        role="col"
      >
        Счет
      </div>
      <div
        class="css-1kiydj1"
        role="col"
      >
        Место
      </div>
    </div>
    <div
      class="css-jqijda"
      role="row"
    >
      <div
        class="css-1hygdxi"
        role="col"
      >
        09.03.2022, 23:58:32
      </div>
      <div
        class="css-182swot"
        role="col"
      >
        Игрок 3
      </div>
      <div
        class="css-lbgr5s"
        role="col"
      >
        300
      </div>
      <div
        class="css-1kiydj1"
        role="col"
      >
        1
      </div>
    </div>
    <div
      class="css-18mgls5"
      role="row"
    >
      <div
        class="css-1hygdxi"
        role="col"
      >
        09.03.2022, 22:51:29
      </div>
      <div
        class="css-182swot"
        role="col"
      >
        Игрок 2
      </div>
      <div
        class="css-lbgr5s"
        role="col"
      >
        200
      </div>
      <div
        class="css-1kiydj1"
        role="col"
      >
        2
      </div>
    </div>
    <div
      class="css-1eso2yd"
      role="row"
    >
      <div
        class="css-1hygdxi"
        role="col"
      >
        09.03.2022, 21:44:21
      </div>
      <div
        class="css-182swot"
        role="col"
      >
        Игрок 1
      </div>
      <div
        class="css-lbgr5s"
        role="col"
      >
        100
      </div>
      <div
        class="css-1kiydj1"
        role="col"
      >
        3
      </div>
    </div>
  </div>
</DocumentFragment>
`;
